lib/SecretsCrypto/sf-crypto-key.c

```diff
@@ -0,0 +1,813 @@
+}
+
+const gchar *sf_crypto_key_get_filter_field(SfCryptoKey *key, const gchar *field)
+{
+	SfCryptoKeyPrivate *priv = sf_crypto_key_get_instance_private(key);
+
+	if (!priv->filter_data)
+		return NULL;
+	return g_hash_table_lookup(priv->filter_data, field);
+}
+
+GVariant *_sf_variant_new_bytes_or_empty(GBytes *bytes)
+{
+	if (!bytes)
+		return g_variant_new_array(G_VARIANT_TYPE_BYTE, NULL, 0);
+	return g_variant_new_fixed_array(G_VARIANT_TYPE_BYTE,
+			g_bytes_get_data(bytes, NULL),
+			g_bytes_get_size(bytes) / sizeof(guchar),
+			sizeof(guchar));
+}
+
+static GVariant *_sf_variant_new_array_bytes_or_empty(GPtrArray *array)
+{
+	GVariantBuilder bldr;
+	size_t i;
+
+	if (!array || array->len == 0)
+		return g_variant_new_array(G_VARIANT_TYPE("ay"), NULL, 0);
+
+	g_variant_builder_init(&bldr, G_VARIANT_TYPE("aay"));
+	for (i = 0; i < array->len; i++)
+		g_variant_builder_add_value(&bldr, _sf_variant_new_bytes_or_empty(g_ptr_array_index(array, i)));
+	return g_variant_builder_end(&bldr);
+}
+
+GBytes *_sf_bytes_new_from_variant_or_null(GVariant *variant)
+{
+	gconstpointer data;
+	gsize n_elements;
+
+	data = g_variant_get_fixed_array(variant, &n_elements, sizeof(guchar));
+	if (n_elements == 0)
+		return NULL;
+	return g_bytes_new(data, n_elements * sizeof(guchar));
+}
+
+static GPtrArray *_sf_array_bytes_from_variant_or_null(GVariant *variant)
+{
+	GVariantIter iter;
+	GVariant *item;
+	GPtrArray *res;
+
+	g_variant_iter_init(&iter, variant);
+
+	if (!(item = g_variant_iter_next_value(&iter)))
+		return NULL;
+	res = g_ptr_array_new_with_free_func((GDestroyNotify)g_bytes_unref);
+	do {
+		g_ptr_array_add(res, _sf_bytes_new_from_variant_or_null(item));
+
+		g_variant_unref(item);
+	} while ((item = g_variant_iter_next_value(&iter)));
+
+	return res;
+}
+
+GVariant *_sf_variant_new_variant_map_or_empty(GHashTable *hash_table)
+{
+	GVariantDict var_dict;
+
+	g_variant_dict_init(&var_dict, NULL);
+
+	if (hash_table) {
+		GHashTableIter ht_iter;
+		gpointer key;
+		gpointer value;
+		g_hash_table_iter_init(&ht_iter, hash_table);
+		while (g_hash_table_iter_next(&ht_iter, &key, &value))
+			g_variant_dict_insert_value(&var_dict, key, value);
+	}
+
+	return g_variant_dict_end(&var_dict);
+}
+
+static GHashTable *_sf_hash_table_new_from_variant(GVariant *variant)
+{
+	GHashTable *res;
+	GVariantIter iter;
+	GVariant *item;
+	gchar *key;
+
+	g_variant_iter_init(&iter, variant);
+
+	if (!g_variant_iter_next(&iter, "{sv}", &key, &item))
+		return NULL;
+
+	res = g_hash_table_new_full(g_str_hash, g_str_equal,
+			g_free, g_free);
+
+	do {
+		g_hash_table_replace(res, key, g_variant_dup_string(item, NULL));
+		g_variant_unref(item);
+	} while (g_variant_iter_next(&iter, "{sv}", &key, &item));
+
+	return res;
+}
+
+GVariant *_sf_crypto_key_to_variant(SfCryptoKey *key)
+{
+	SfCryptoKeyPrivate *priv = sf_crypto_key_get_instance_private(key);
+
+	return g_variant_new("((sss)iiiiiayayaya(ay)a{sv})",
+			priv->name ?: "",
+			priv->collection_name ?: "",
```

Comment:
Seems unlikely, but if we switch to Clang then I'm not sure the elvis operator will continue to work, I think it might be gnu-extension